File: python/tests/test_rag.py
"""Tests for src/rag.py — Week 3"""
import pytest
from src.rag import (
    index_documents, retrieve, hybrid_search,
    grounded_answer, grounded_answer_with_chunks,
)


@pytest.fixture(autouse=True)
def _build_index():
    """Index the doc set fresh before each test."""
    import src.rag as rag
    rag._vectorstore = None
    rag._documents = None
    rag._embeddings = None
    count = index_documents(chunk_size=512, chunk_overlap=64)
    assert count > 0, "No documents were indexed"


def test_rag_imports_llm():
    """rag.py imports from llm.py — the import graph holds."""
    from src.rag import grounded_answer
    import inspect
    source = inspect.getsource(grounded_answer)
    assert "get_llm" in source, "grounded_answer should call get_llm() from src.llm"


def test_index_creates_chunks():
    """Indexing produces chunks stored in ChromaDB."""
    # Re-index to get fresh count
    count = index_documents(chunk_size=512)
    assert count >= 4, f"Expected at least 4 chunks, got {count}"


def test_retrieve_returns_relevant_chunks():
    """Retrieval finds chunks relevant to the query."""
    chunks = retrieve("What endpoints does the payment API expose?", k=3)
    assert len(chunks) > 0, "No chunks retrieved"
    assert any("payment" in c.lower() for c in chunks), (
        "Retrieved chunks should mention 'payment'"
    )


def test_retrieve_returns_different_chunks_for_different_queries():
    """Different queries return different chunks."""
    pay_chunks = retrieve("payment API timeout", k=3)
    contrib_chunks = retrieve("how to contribute", k=3)
    assert pay_chunks != contrib_chunks, (
        "Different queries should return different chunks"
    )


def test_grounded_answer_uses_context():
    """The LLM answers from the retrieved context, not training data."""
    answer = grounded_answer("What is the SLA for the payment API?", k=3)
    assert len(answer) > 10, "Answer is too short"
    # Should reference the SLA from the payment API spec
    assert any(
        phrase in answer.lower()
        for phrase in ["99.95", "sla", "uptime"]
    ), f"Answer doesn't reference the payment API SLA: {answer[:100]}"


def test_grounded_answer_with_chunks_returns_both():
    """grounded_answer_with_chunks returns answer + chunks."""
    answer, chunks = grounded_answer_with_chunks(
        "What is the SLA for the payment API?", k=3
    )
    assert len(answer) > 10
    assert len(chunks) > 0
    assert any("99.95" in c or "SLA" in c for c in chunks), (
        "Chunks should contain SLA information"
    )


def test_out_of_corpus_question_says_no_information():
    """An out-of-corpus question should not hallucinate."""
    answer = grounded_answer(
        "What's the revenue forecast for Q4 2028?", k=3
    )
    assert (
        "don't have" in answer.lower()
        or "no information" in answer.lower()
    ), f"Model should decline to answer, got: {answer[:100]}"


def test_hybrid_search_includes_keyword_matches():
    """Hybrid search should return results for keyword queries."""
    hyb = hybrid_search("payment-api", k=3)
    vec = retrieve("payment-api", k=3)
    assert len(hyb) > 0, "Hybrid search returned no results"
    # Hybrid should at minimum match vector quality
    assert len(hyb) == len(vec), (
        f"Hybrid ({len(hyb)}) should return same count as vector ({len(vec)})"
    )


def test_chunk_size_changes_chunk_count():
    """Different chunk sizes produce different numbers of chunks."""
    count_256 = index_documents(chunk_size=256)
    count_1024 = index_documents(chunk_size=1024)

    # Smaller chunks = more chunks. Larger chunks = fewer.
    assert count_256 > count_1024, (
        f"Expected chunk_size=256 ({count_256}) to produce more chunks "
        f"than chunk_size=1024 ({count_1024})"
    )
    # Restore default size for subsequent tests
    index_documents(chunk_size=512)


def test_unindexed_retrieve_raises():
    """Calling retrieve without an index raises a clear error."""
    import src.rag as rag
    old_vs = rag._vectorstore
    rag._vectorstore = None
    with pytest.raises(RuntimeError, match="No index found"):
        retrieve("test query")
    rag._vectorstore = old_vs

File: nodejs/tests/test_rag.js
/**
 * Tests for src/rag.js — Week 3 (Node.js)
 */
import { describe, it, expect, beforeAll } from "vitest";
import {
  indexDocuments,
  retrieve,
  retrieveWithSources,
  hybridSearch,
  hybridSearchWithScores,
  groundedAnswer,
  groundedAnswerWithChunks,
  _bm25Tokenize,
  _isTitleOnly,
  SYSTEM_PROMPT,
  RRF_K,
} from "../src/rag.js";

// Build index once before all tests
beforeAll(async () => {
  const count = await indexDocuments(null, 512, 64);
  expect(count).toBeGreaterThan(0);
});

describe("indexDocuments", () => {
  it("creates at least 4 chunks", async () => {
    const count = await indexDocuments(null, 512, 64);
    expect(count).toBeGreaterThanOrEqual(4);
  });
});

describe("retrieve", () => {
  it("returns relevant chunks for a payment API query", async () => {
    const chunks = await retrieve(
      "What endpoints does the payment API expose?",
      3
    );
    expect(chunks.length).toBeGreaterThan(0);
    const hasPayment = chunks.some((c) =>
      c.toLowerCase().includes("payment")
    );
    expect(hasPayment).toBe(true);
  });

  it("returns different chunks for different queries", async () => {
    const payChunks = await retrieve("payment API timeout", 3);
    const contribChunks = await retrieve("how to contribute", 3);
    // Should differ since queries target different topics
    const payJoined = payChunks.join("");
    const contribJoined = contribChunks.join("");
    expect(payJoined).not.toBe(contribJoined);
  });
});

describe("groundedAnswer", () => {
  it("answers from retrieved context, not training data", async () => {
    const answer = await groundedAnswer(
      "What is the SLA for the payment API?",
      3
    );
    expect(answer.length).toBeGreaterThan(10);
    const lower = answer.toLowerCase();
    expect(
      lower.includes("99.95") ||
        lower.includes("sla") ||
        lower.includes("uptime")
    ).toBe(true);
  });

  it("returns answer and chunks", async () => {
    const [answer, chunks] = await groundedAnswerWithChunks(
      "What is the SLA for the payment API?",
      3
    );
    expect(answer.length).toBeGreaterThan(10);
    expect(chunks.length).toBeGreaterThan(0);
    expect(
      chunks.some((c) => c.includes("99.95") || c.includes("SLA"))
    ).toBe(true);
  });

  it("does not hallucinate on out-of-corpus questions", async () => {
    const answer = await groundedAnswer(
      "What's the revenue forecast for Q4 2028?",
      3
    );
    const lower = answer.toLowerCase();
    expect(
      lower.includes("don't have") || lower.includes("no information")
    ).toBe(true);
  });
});

describe("hybridSearch", () => {
  it("returns results for keyword queries", async () => {
    const hyb = await hybridSearch("payment-api", 3);
    const vec = await retrieve("payment-api", 3);
    expect(hyb.length).toBeGreaterThan(0);
    expect(hyb.length).toBe(vec.length);
  });
});

describe("chunk size", () => {
  it("smaller chunk size produces more chunks", async () => {
    const count256 = await indexDocuments(null, 256, 64);
    const count1024 = await indexDocuments(null, 1024, 64);
    expect(count256).toBeGreaterThan(count1024);
    // Restore default
    await indexDocuments(null, 512, 64);
  });
});

describe("error handling", () => {
  it("throws when retrieve is called without an index", async () => {
    // We can't easily reset the module-level state in ESM,
    // but the error message is clear in the code.
    // This test just validates the function exists and is importable.
    expect(typeof retrieve).toBe("function");
  });
});

// ═══════════════════════════════════════════════════════════════
// Week 4 additions — provenance, hybrid ranks, tokenizer
// ═══════════════════════════════════════════════════════════════

describe("retrieveWithSources", () => {
  it("returns content, source, and score", async () => {
    const results = await retrieveWithSources("payment API", 3);
    expect(results.length).toBeGreaterThan(0);
    expect(results[0].content).toBeTruthy();
    expect(results[0].source).toBeTruthy();
    expect(typeof results[0].score).toBe("number");
  });
});

describe("hybridSearchWithScores", () => {
  it("reports per-retriever ranks and the fused score", async () => {
    const results = await hybridSearchWithScores("payment-api", 3);
    expect(results.length).toBeGreaterThan(0);
    expect(results[0]).toHaveProperty("rrf_score");
    expect(results[0]).toHaveProperty("vec_rank");
    expect(results[0]).toHaveProperty("bm25_rank");
    expect(results[0].rrf_score).toBeGreaterThan(0);
  });
});

describe("BM25 tokenizer & title-only filter", () => {
  it("splits hyphenated IDs into tokens", () => {
    expect(_bm25Tokenize("INC-799")).toEqual(["inc", "799"]);
  });

  it("detects heading-only chunks", () => {
    expect(_isTitleOnly("# Payment API\n> owner: x")).toBe(true);
    expect(
      _isTitleOnly(
        "# Payment API\nThis service handles payments and refunds with retries."
      )
    ).toBe(false);
  });

  it("exposes the RRF constant and a context prompt slot", () => {
    expect(RRF_K).toBe(60);
    expect(SYSTEM_PROMPT).toContain("{context}");
  });
});
